orangecontrib/prototypes/widgets/owimagegrid.py

```diff
@@ -0,0 +1,1135 @@
+            item.setThumbnailSize(QSizeF(item_size, item_size))
+
+        self.__relayoutGrid(self.__columnCount)
+
+    def __relayoutGrid(self, columnCount):
+        layout = self.__layout
+        if columnCount == layout.columnCount():
+            return
+
+        # remove all items from the layout, then re-add them back in
+        # updated positions
+        items = self.__takeItemsFrom(0)
+        for i, item in enumerate(items):
+            layout.addItem(item, i // columnCount, i % columnCount)
+
+    def items(self):
+        """
+        Return all thumbnail items.
+
+        Returns
+        -------
+        thumbnails : List[GraphicsThumbnailWidget]
+        """
+        return list(self.__thumbnails)
+
+    def currentItem(self):
+        """
+        Return the current (last focused) thumbnail item.
+        """
+        return self.__current
+
+    def _fitncols(self, widths, spacing, constraint):
+        def sliced(seq, ncol):
+            return [seq[i:i + ncol] for i in range(0, len(seq), ncol)]
+
+        def flow_width(widths, spacing, ncol):
+            W = sliced(widths, ncol)
+            col_widths = map(max, zip_longest(*W, fillvalue=0))
+            return sum(col_widths) + (ncol - 1) * spacing
+
+        ncol_best = 1
+        for ncol in range(2, len(widths) + 1):
+            w = flow_width(widths, spacing, ncol)
+            if w <= constraint:
+                ncol_best = ncol
+            else:
+                break
+
+        return ncol_best
+
+    def keyPressEvent(self, event):
+        if event.key() in [Qt.Key_Left, Qt.Key_Right, Qt.Key_Up, Qt.Key_Down]:
+            self._moveCurrent(event.key(), event.modifiers())
+            event.accept()
+            return
+        super().keyPressEvent(event)
+
+    def eventFilter(self, receiver, event):
+        if isinstance(receiver, GraphicsThumbnailWidget) and \
+                        event.type() == QEvent.FocusIn and \
+                        receiver in self.__thumbnails:
+            self.__current = receiver
+            self.currentThumbnailChanged.emit(receiver)
+
+        return super().eventFilter(receiver, event)
+
+    def _moveCurrent(self, key, modifiers=Qt.NoModifier):
+        """
+        Move the current thumbnail focus (`currentItem`) based on a key press
+        (Qt.Key{Up,Down,Left,Right})
+
+        Parameters
+        ----------
+        key : Qt.Key
+        modifiers : Qt.Modifiers
+        """
+        current = self.__current
+        layout = self.__layout
+        columns = layout.columnCount()
+        rows = layout.rowCount()
+        itempos = {}
+        for i, j in itertools.product(range(rows), range(columns)):
+            if i * columns + j >= layout.count():
+                break
+            item = layout.itemAt(i, j)
+            if item is not None:
+                itempos[item] = (i, j)
+        pos = itempos.get(current, None)
+
+        if pos is None:
+            return False
+
+        i, j = pos
```

Comment:
This code above, is it the same as this?
```py
for i in range(layout.rowCount()):
    for j in range(layout.columnCount()):
        if layout.itemAt(i, j) == current:
            break
else:
    return False

...  # do something with i and j
```

Reply:
I see it's copied line for line from Image Viewer. Never mind. If it's good enough for main ...